### PR TITLE
CRIU removes libj9criu29

### PR DESCRIPTION
**Pull request by JasonFengJ9**
CRIU removes `libj9criu29`

This library is not needed after all its functionalities are moved into `java.base`.

Verified via [a personal build](https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/20155/) 

Depend on 
* https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/725
* https://github.com/ibmruntimes/openj9-openjdk-jdk22/pull/10
* https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/114
* https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/306
* https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/749
* https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/729

Signed-off-by: Jason Feng <fengj@ca.ibm.com>

### CODE REVIEW

**Comment by pshipton** [comment]
I'm setting this as draft just because I want to get the latest OpenJDK changes into 0.43 before we diverge the extensions.

File: jcl/src/java.base/share/classes/openj9/internal/criu/InternalCRIUSupport.java [hunk 1]
```diff
@@ -202,35 +199,6 @@ public static enum HookMode {
 		CONCURRENT_MODE
 	}
 
-	private static boolean loadNativeLibrary() {
-		if (!nativeLoaded) {
-			try {
-				AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
-					System.loadLibrary("j9criu29"); //$NON-NLS-1$
```
Comment (tajila) [inline jcl/src/java.base/share/classes/openj9/internal/criu/InternalCRIUSupport.java:209]: `System.loadLibrary("j9criu29"); ` is not needed, but we still need to call `setupJNIFieldIDsAndCRIUAPI()` to do a dynamic load of criu at class init so we know before hand if there are issues with the library when calling `isCRIUSupportEnabled`

Comment (tajila) [reply jcl/src/java.base/share/classes/openj9/internal/criu/InternalCRIUSupport.java:209]: I think some of the checks were lost after doing the refactoring as nothing calls `isCRIUSupportEnabledAndNativeLoaded` anymore

File: jcl/src/java.base/share/classes/openj9/internal/criu/InternalCRIUSupport.java [hunk 2]
```diff
@@ -269,34 +237,15 @@ public InternalCRIUSupport(Path imageDir) {
...
-
 	/**
 	 * Returns an error message describing why isCRIUSupportEnabled()
 	 * returns false, and what can be done to remediate the issue.
 	 *
-	 * @return NULL if isCRIUSupportEnabled() returns true and nativeLoaded is true as well, otherwise the error message.
+	 * @return NULL if isCRIUSupportEnabled() returns true, otherwise the error message.
 	 */
 	public static String getErrorMessage() {
-		if (errorMsg == null) {
```
Comment (tajila) [inline jcl/src/java.base/share/classes/openj9/internal/criu/InternalCRIUSupport.java:289]: all this stuff needs to remain

File: jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java [hunk 3]
```diff
@@ -93,7 +93,7 @@ public CRIUSupport(Path imageDir) {
 	}
 
 	/**
-	 * Queries if CRIU support is enabled and j9criu29 library has been loaded.
+	 * Queries if CRIU support is enabled.
```
Comment (tajila) [inline jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java:96]: this needs to call `isCRIUSupportEnabledAndNativeLoaded`

File: jcl/src/java.base/share/classes/openj9/internal/criu/InternalCRIUSupport.java (hunk 1 above)
Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/openj9/internal/criu/InternalCRIUSupport.java:209]: Added back `setupJNIFieldIDsAndCRIUAPI()` along with `loadNativeLibrary()` & `init()`.
Invoke it from `InternalCRIUSupport.checkpointJVM()` and `CRIUSupport.isCRIUSupportEnabled()`.

File: jcl/src/java.base/share/classes/openj9/internal/criu/InternalCRIUSupport.java (hunk 2 above)
Comment (JasonFengJ9) [reply jcl/src/java.base/share/classes/openj9/internal/criu/InternalCRIUSupport.java]: Added it back.

File: jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java (hunk 3 above)
Comment (JasonFengJ9) [reply jcl/src/openj9.criu/share/classes/org/eclipse/openj9/criu/CRIUSupport.java]: `isCRIUSupportEnabledAndNativeLoaded()` call is added.

**Comment by JasonFengJ9** [comment]
https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/123 is merged.
Moving this out of draft state.

File: jcl/src/java.base/share/classes/openj9/internal/criu/InternalCRIUSupport.java [hunk 4]
```diff
@@ -868,7 +856,7 @@ private static void clearInetAddressCache() {
 	 *                                       restore
 	 */
 	public synchronized void checkpointJVM() {
-		if (isCRIUSupportEnabled()) {
+		if (isCRIUSupportEnabledAndNativeLoaded()) {
```
Comment (tajila) [inline jcl/src/java.base/share/classes/openj9/internal/criu/InternalCRIUSupport.java:859]: you can leave this as is, as it will be changed in the next PR

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk21

**Comment by tajila** [comment]
jenkins test sanity zlinux jdk17

**Comment by pshipton** [comment]
> I think we should get this into 0.43 if possible

Is it worth a delay to the release?

**Comment by pshipton** [comment]
Are there any security aspects to having the library available when it's not used?

**Comment by tajila** [comment]
> Is it worth a delay to the release?

No

> Are there any security aspects to having the library available when it's not used?

Im only concerned about the missing check for the native library load in `CRIUSupport::isCRIRSupportEnabled`

**Comment by JasonFengJ9** [comment]
> Build_JDK21_aarch64_linux_Personal — Build FAILED

`17:37:53  curl: (56) OpenSSL SSL_read: Connection reset by peer, errno 104`
Infra error

**Comment by pshipton** [comment]
As I understand it there is a library containing the following natives but no equivalent Java code, so any app could add the java code and call these natives.

Java_org_eclipse_openj9_criu_CRIUSupport_checkpointJVMImpl
Java_org_eclipse_openj9_criu_CRIUSupport_getRestoreSystemProperites
Java_org_eclipse_openj9_criu_CRIUSupport_setupJNIFieldIDsAndCRIUAPI

**Comment by pshipton** [comment]
No RC1 builds have been started, and I need to update an OpenJ9 change to cover jdk21 so there is some time to get this in.

**Comment by tajila** [comment]
jenkins test sanity alinux64 jdk21

**Comment by JasonFengJ9** [comment]
> 19:17:27  gmake[3]: *** No rule to make target '/home/jenkins/workspace/Build_JDK17_s390x_linux_Personal/build/linux-s390x-server-release/vm/runtime/libj9criu29.so', needed by '/home/jenkins/workspace/Build_JDK17_s390x_linux_Personal/build/linux-s390x-server-release/support/modules_libs/openj9.criu/default/libj9criu29.so'.  Stop.

This PR depends on 
* https://github.com/ibmruntimes/openj9-openjdk-jdk/pull/725
* https://github.com/ibmruntimes/openj9-openjdk-jdk22/pull/10
* https://github.com/ibmruntimes/openj9-openjdk-jdk21/pull/114
* https://github.com/ibmruntimes/openj9-openjdk-jdk17/pull/306
* https://github.com/ibmruntimes/openj9-openjdk-jdk11/pull/749
* https://github.com/ibmruntimes/openj9-openjdk-jdk8/pull/729

**Comment by pshipton** [comment]
jenkins test sanity alinux64 jdk21 depends ibmruntimes/openj9-openjdk-jdk21#114

**Comment by pshipton** [comment]
jenkins test sanity zlinux jdk17 depends ibmruntimes/openj9-openjdk-jdk17#306

**Comment by JasonFengJ9** [comment]
> Im only concerned about the missing check for the native library load in CRIUSupport::isCRIRSupportEnabled

Just realized that `0.43` doesn't have
* https://github.com/eclipse-openj9/openj9/pull/18417
which missed the check in question.

That PR was to prepare the adoption of `jdk.crac` 
* https://github.com/eclipse-openj9/openj9/pull/18627
still open.

Since the issue wasn't in `0.43`, this PR isn't needed there either.

**Comment by tajila** [comment]
> Since the issue wasn't in 0.43, this PR isn't needed there either.

okay, thats good

**Pull request merged**